srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,1571 @@
+//!
+//! The election happens in _rounds_: every `N` blocks, all previous members are retired and a new
+//! set is elected (which may or may not have an intersection with the previous set). Each round
+//! lasts for some number of blocks defined by `TermDuration` storage item. The words _term_ and
+//! _round_ can be used interchangeably in this context.
+//!
+//! `TermDuration` might change during a round. This can shorten or extend the length of the round.
+//! The next election round's block number is never stored but rather always checked on the fly.
+//! Based on the current block number and `TermDuration`, the condition `BlockNumber % TermDuration
+//! == 0` being satisfied will always trigger a new election round.
+//!
+//! ### Voting
+//!
+//! Voters can vote for any set of the candidates by providing a list of account ids. Invalid votes
+//! (voting for non-candidates) are ignored during election. Yet, a voter _might_ vote for a future
+//! candidate. Voters reserve a bond as they vote. Each vote defines a `value`. This amount is
+//! locked from the account of the voter and indicates the weight of the vote. Voters can update
+//! their votes at any time by calling `vote()` again. This keeps the bond untouched but can
+//! optionally change the locked `value`. After a round, votes are kept and might still be valid for
+//! further rounds. A voter is responsible for calling `remove_voter` once they are done to have
+//! their bond back and remove the lock.
+//!
+//! Voters also report other voters as being defunct to earn their bond. A voter is defunct once all
+//! of the candidates that they have voted for are neither a valid candidate anymore nor a member.
+//! Upon reporting, if the target voter is actually defunct, the reporter will be rewarded by the
+//! voting bond of the target. The target will lose their bond and get removed. If the target is not
+//! defunct, the reporter is slashed and removed. To prevent being reported, voters should manually
+//! submit a `remove_voter()` as soon as they are in the defunct state.
+//!
+//! ### Candidacy and Members
+//!
+//! Candidates also reserve a bond as they submit candidacy. A candidate cannot take their candidacy
+//! back. A candidate can end up in one of the below situations:
+//!   - **Winner**: A winner is kept as a _member_. They must still have a bond in reserve and they
+//!     are automatically counted as a candidate for the next election.
+//!   - **Loser**: Any of the candidate who are not a winner are left as losers. A loser might be an
+//!     _outgoing member_, meaning that they are an active member who failed to keep their spot. In
+//!     this case, the outgoing member will get their bond back. Otherwise, the bond is slashed from
+//!     the loser candidate.
+//!   - **Runner-up**: Runner ups are the best candidates immediately after the winners. The number
+//!     of runner-ups to keep is configurable. Runner-ups serve one major role: if a member is
```

Comment:
```suggestion
//!     of runner-ups to keep is configurable. Runners-up serve one major role: if a member is
```

The main point of runners-up is not to provide replacements for removed candidates (indeed, this might be a security risk). Rather it's to ensure that the number of candidates is kept fairly low by periodically culling the crappy candidates (and burning their deposit to disincentivise frivolous usage).

Reply:
My writing here (everything after: _Runner-ups serve one __major__ role:..._ should be gone) is too constrained, I agree, but I don't also quite get what you mean by:

> Rather it's to ensure that the number of __candidates__ is kept fairly low by periodically culling the crappy candidates

I don't see how the runner-ups limit the number of candidates; candidates always strat from an empty list at the beginning of a term and are always thrown away (and forfeit bond) after election happens.